File: byteback-test/test-java-8/src/main/java/byteback/test/algorithm/ArrayReverse.java
/**
 * RUN: %{byteback} -cp %{jar} -c %{class} -o %t.bpl
 */
package byteback.test.algorithm;

import static byteback.annotations.Contract.*;
import static byteback.annotations.Operator.*;
import static byteback.annotations.Quantifier.*;
import static byteback.annotations.Special.*;

import javax.annotation.processing.SupportedSourceVersion;

import byteback.annotations.Binding;

public class ArrayReverse {

	@Pure
	public static boolean reverse_of(final int[] a, final int[] b) {
		int i = Binding.integer();

		return and(eq(a.length, b.length),
							 forall(i, implies(lte(0, i) & lt(i, a.length),
																 eq(a[i], b[b.length - 1 - i]))));
	}

	@Pure
	@Predicate
	public static boolean bounded_index(final int[] a, final int i) {
		return lte(0, i) & lt(i, a.length);
	}

	@Pure
	@Predicate
	public static boolean bounded_indices(final int[] a, final int i, final int j) {
		return bounded_index(a, i) & bounded_index(a, j);
	}

	@Predicate
	public static boolean swapped_elements(final int[] a, final int i, final int j) {
		return eq(old(a[i]), a[j]) & eq(old(a[j]), a[i]);
	}

	@Predicate
	public static boolean invariant_elements(final int[] a, final int i, final int j) {
		int m = Binding.integer();

		return forall(m, implies(lte(0, m) & lt(m, i) | lt(j, m) & lt(m, a.length), eq(a[i], old(a[i]))));
	}

	@Return
	@Require("bounded_indices")
	@Ensure("swapped_elements")
	@Ensure("invariant_elements")
	public static void swap(final int[] a, int i, int j) {
		final int y = a[i];
		a[i] = a[j];
		a[j] = y;
	}

	@Pure
	@Predicate
	public static boolean array_is_null(int[] a) {
		return eq(a, null);
	}

	@Predicate
	public static boolean reversed(int[] a) {
		return implies(not(array_is_null(a)), reverse_of(a, old(a)));
	}

	@Raise(exception = IllegalArgumentException.class, when = "array_is_null")
	@Ensure("reversed")
	public static void reverse(int[] a) {
		if (a == null) {
			throw new IllegalArgumentException("Input array cannot be null");
		}

		final int l = a.length - 1;
		int i = 0;

		while (i < (l - i)) {
			invariant(lte(0, i) & lte(i, (l + 1) / 2));
			swap(a, i, l - i);
		}
	}

}

/**
 * RUN: %{verify} %t.bpl | filecheck %s
 * CHECK: Boogie program verifier finished with 3 verified, 0 errors
 */

File: byteback-test/test-kotlin-1.8.0/src/main/kotlin/byteback/test/algorithms/ArrayReverse.kt
/**
 * RUN: %{byteback} -cp %{jar} -c %{class} -o %t.bpl
 */
package byteback.test.algorithms;

import byteback.annotations.Contract.*
import byteback.annotations.Operator.*
import byteback.annotations.Special.*
import byteback.annotations.Quantifier.*
import byteback.annotations.Binding

class ArrayReverse {

		@Pure
		fun `reverse of`(a: IntArray?, b: IntArray?): Boolean {
				val i: Int = Binding.integer();

				return and(eq(a!!.size, b!!.size),
									 forall(i, implies(lte(0, i) and lt(i, a.size), eq(a[i], b[b.size - 1 - i]))))
		}

		@Pure
		@Predicate
		fun `array is not null`(a: IntArray?, i: Int, j: Int): Boolean {
				return neq(a, null);
		}

		@Predicate
		fun `swapped elements`(a: IntArray?, i: Int, j: Int): Boolean {
				return eq(old(a!![i]), a[j]) and eq(old(a[j]), a[i]);
		}

		@Return(`when` = "array is not null")
		fun swap(a: IntArray, i: Int, j: Int) {
				val y: Int = a[i]
				a[i] = a[j];
				a[j] = y;
		}

		@Predicate
		fun `reverses the array`(a: IntArray?): Boolean {
				return implies(not(`array is null`(a)), `reverse of`(a, old(a)));
		}

		@Pure
		@Predicate
		fun `array is null`(a: IntArray?): Boolean {
				return eq(a, null);
		}

		@Raise(`exception` = IllegalArgumentException::class, `when` = "array is null")
		@Ensure("reverses the array")
		fun reverse(a: IntArray) {
				val l = a.size - 1
				var i = 0

				while (i < (l - i)) {
						invariant(lte(0, i) and lte(i, (l + 1) / 2));
						swap(a, i, l - i);
				}
		}

}

/**
 * RUN: %{verify} %t.bpl
 */
